src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,837 @@
+        for expr in self.iter_expr_mut() {
+            expr.reduce(relation_type);
+        }
+    }
+
+    /// Maps all the expressions within the Constraint.
+    fn map_to_global(mut self, input_mapper: &JoinInputMapper, index: usize) -> Self {
+        for expr in self.iter_expr_mut() {
+            *expr = input_mapper.map_expr_to_global(expr.to_owned(), index);
+        }
+        self
+    }
+
+    /// Attempts to perform substitutions via `map` in all the expressions contained
+    /// in the constraint and returns `false` if the constraint cannot be lifted.
+    fn substitute(&mut self, map: &HashMap<MirScalarExpr, MirScalarExpr>) -> bool {
+        match self {
+            Self::Equivalence(class) => {
+                // Keep the class members that can find substitutions
+                *class = class
+                    .drain(..)
+                    .flat_map(|mut expr| {
+                        if substitute(&mut expr, map) {
+                            Some(expr)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                class.len() > 1
+            }
+            Self::Predicate(p) => substitute(p, map),
+        }
+    }
+
+    /// Optimizes all the expressions in the constraint by performing replacements.
+    /// Returns true if any replacement took place.
+    fn optimize(&mut self, predicates: &PredicateStructure) -> bool {
+        self.iter_expr_mut()
+            .map(|e| optimize(e, predicates))
+            .fold(false, |acc, v| acc || v)
+    }
+}
+
+impl std::fmt::Display for Constraint {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        match self {
+            Self::Equivalence(class) => {
+                write!(f, "Equivalence[")?;
+                for (index, expr) in class.iter().enumerate() {
+                    write!(f, "{}{}", if index > 0 { " " } else { "" }, expr)?;
+                }
+                write!(f, "]")
+            }
+            Self::Predicate(p) => p.fmt(f),
+        }
+    }
+}
+
+/// Adds a constraint, expressed as a predicate, to the collection.
+///
+/// Equality predicates are converted into an equivalence and an IS NOT NULL predicate.
+fn append_constraint(constraints: &mut Vec<Constraint>, predicate: MirScalarExpr) {
+    if let MirScalarExpr::CallBinary {
+        expr1,
+        expr2,
+        func: BinaryFunc::Eq,
+    } = predicate
+    {
+        constraints.push(Constraint::Predicate(
+            expr1
+                .clone()
+                .call_unary(UnaryFunc::IsNull(expr::func::IsNull))
+                .call_unary(UnaryFunc::Not(expr::func::Not)),
+        ));
+        constraints.push(Constraint::Equivalence(vec![*expr1, *expr2]));
+    } else {
+        constraints.push(Constraint::Predicate(predicate));
+    }
+}
+
+/// Denormalizes the predicates in the given list for easily computing the
```

Comment:
Good points.

The information we can lift through a union is the intersection of the information lifted from all its branches. If we had a union of two branches where from each branch we know that:

```
branch 1: [Equivalence[#0 #1 #2], Predicate(#0 > 10)]
branch 2: [Equivalence[#1 #2 #3], Predicate(#1 > 10)]
intersection: []
```

If we intersected these two lists of constraints without denormalizing them we would get an empty list. However, if we denormalized them, we get:

```
denormalized constraints from branch 1: [Equivalence[#0 #1], Equivalence[#0 #2], Equivalence[#1 #2], Predicate(#0 > 10), Predicate(#1 > 10), Predicate(#2 > 10)]
denormalized constraints from  branch 2: [Equivalence[#1 #2], Equivalence[#1#3], Equivalence[#2 #3], Predicate(#1 > 10), Predicate(#2 > 10), Predicate(#3 > 10)]
intersection: [Equivalence[#1 #2], Predicate(#1 > 10), Predicate(#2 > 10)]
```

which once normalized again results in:

```
normalized intersection: [Equivalence[#1 #2], Predicate(#1 > 10)]
```

We could even try something smarter and run some range analysis so that we end up lifting things like:

```
branch 1: [Equivalence[#0 #1 #2], Predicate(#0 > 10)]
branch 2: [Equivalence[#1 #2 #3], Predicate(#1 > 20)]
normalized intersection: [Equivalence[#1 #2], Predicate(#1 > 10)]
```

but that's for a follow-up task.

Reply:
Couldn't we do it over normalized predicates as follows:

Let branch 1 and branch 2 contain a set of equivalences and a set of predicates each.

```
branch 1: { EQ_1, PR_1 } 
branch 2: { EQ_2, PR_2 }
```

Then the output set of equivalences is given by

```
EQ_O = { eq_1 ∩ eq_2  | eq_1 ∈ Eq_1, eq_2 ∈ EQ_2 }
```

where the set of equivalent classes `EQ_R` can be additionally normalized by merging members that have a non-empty intersection (for example the equivalence classes `{1, 2}` and `{2, 3}` can be merged into `{1, 2, 3}`).

The resulting set of predicates can be computed as follows:

1. Uniquely determine a representative `r` from each equivalence class `eq` in `EQ_O` (as you suggested: a constant, the reference with the highest ID, or ... ???),
2. In each predicate `pr ∈ PR_O`, replace references to members of `eq` with `r`.
3. Remove duplicates from `PR_O`.

I am not sure how to handle the `???` part, but for your example (which does not have this type of complexity), I think the calculations lead to the same result. FYI, this will not be a problem if our expressions were in some sort of SSA form.

Reply:
Sure, it can be done that way too. We would need rewrite this:

```
                MirRelationExpr::Union { base, inputs } => {
                    let mut know1 = denormalize_predicates(self.action(base, let_knowledge)?);
                    for input in inputs.iter_mut() {
                        let know2 = denormalize_predicates(self.action(input, let_knowledge)?);
                        know1.retain(|predicate| know2.contains(predicate));
                    }
                    know1
                }
```

as something like:

```
                MirRelationExpr::Union { base, inputs } => {
                    let (mut equivalences1, mut predicates1) =
                        split_equivalences(self.action(base, let_knowledge)?);
                    for input in inputs.iter_mut() {
                        let (mut equivalences2, mut predicates2) =
                            split_predicates(self.action(input, let_knowledge)?);
                        equivalences1 = merge_equivalences(equivalences1, equivalences2);
                        normalize_predicates_using_equivalences(&mut predicates1, &equivalences1);
                        normalize_predicates_using_equivalences(&mut predicates2, &equivalences1);
                        predicates1.retain(|predicate| predicates2.contains(predicate));
                    }
                    equivalences
                        .into_iter()
                        .map(|e| Constraint::Equivalence(e))
                        .chain(predicates.into_iter().map(|p| Constraint::Predicate(p)))
                }

```

However, `denormalize_predicates` is also useful, and used for, propagating information through operators that may drop some columns as it happens with `Project` and `Reduce` operators. See this [commit](https://github.com/MaterializeInc/materialize/pull/9354/commits/83b0b0def88ccaad413b5d0b1ff137f7113e0019) and the tests included on it. For example, when a `Project` operators drops the column that is the representative of a class we still want to propagate what we know about the survival members of that class. Before that commit we were just losing the information if the representative was projected out.

Reply:
I should extend the comment in `denormalize_predicates` to include all the cases where it is used.